### PR TITLE
treewide: mass rebuild complicated cleanups

### PR DESCRIPTION
**Pull request by oxij**
###### Motivation for this change

Continuation of #47233.

The first two patches homogenize gcc expressions (i.e. make `diff` between them as small as possible) in the hope that they could be merged into a single expression eventually.

I have no idea whether the third patch does the right thing.

###### Things done

- [X] Running on top of this for two months without issues.

/cc @Ericson2314

### CODE REVIEW

**Comment by oxij** [comment]
I see. Done.

File: pkgs/development/compilers/gcc/4.9/default.nix
Comment (veprbl) [inline pkgs/development/compilers/gcc/4.9/default.nix]: IIRC gcc 4.8 is broken on darwin. Perhaps, the same is true for 4.9. Should probably use the ```meta.broken``` attribute to indicate that.

File: pkgs/development/compilers/gcc/5/default.nix
Comment (veprbl) [inline pkgs/development/compilers/gcc/5/default.nix]: I think you missed one on line 238 here.

**Comment by FRidh** [comment]
Waiting for more feedback before merging into staging.

**Comment by oxij** [comment]
@veprbl

> I think you missed one on line 238 here.

Oops, right, fixed.

> IIRC gcc 4.8 is broken on darwin. Perhaps, the same is true for 4.9. Should probably use the ```meta.broken``` attribute to indicate that.

But the expression for 4.8 has `darwin` in platforms and I didn't touch that part.

**Comment by oxij** [comment]
ping?

All of this builds fine on `x86_64-linux` and `meta` nitpicks can be fixed separately, IMHO.

**Review by 7c6f434c — APPROVED** [review]
If there is an explicit fresh review request towards @xeji I want to allow some time for it to happen (or not). Ping me in 24 hours, I will merge provided @FRidh doesn't merge it by then (assuming @xeji doesn't have hash-affecting objections, of course)

**Comment by xeji** [comment]
I don't feel competent to review this since I'm not familiar with the details of our gcc and llvm expressions.

**Comment by 7c6f434c** [comment]
I am not keeping a close track on the minor details, but I can understand what these changes do, and I agree these are a mix of useful cleanups and neutral layout changes.

File: pkgs/development/compilers/llvm/5/llvm.nix
Comment (Ericson2314) [inline pkgs/development/compilers/llvm/5/llvm.nix]: Oh for the record this is because `compiler-rt` is now built separately and `libcxxabi` was just needed for that. Overall on the LLVM side, tools and runtime libraries are never built in the same derivation, which is good.

**Pull request merged**
